Tools/clinic/clinic.py

```diff
@@ -2682,9 +2682,18 @@ def displayname(self) -> str:
 
     @functools.cached_property
     def fulldisplayname(self) -> str:
-        if isinstance(self.module, Module):
-            return f"{self.module.name}.{self.displayname}"
-        return self.displayname
+        name = self.displayname
+        try:
+            if self.kind.new_or_init:
+                parent = self.cls.parent
+            else:
+                parent = self.parent
+            while parent:
+                name = f"{parent.name}.{name}"
+                parent = parent.parent
```

Comment:
@AlexWaygood: uh-oh, mypy seems to hate this approach :(

Reply:
An alternative could be to make it recursive. Should make the typing slightly easier.

Reply:
Ohh, I see what you're doing here. Yeah, type checkers _are_ generally going to push you towards LBYL styles instead of EAFP styles for this kind of thing, I'm afraid ://

Here's a rewrite that's only one line longer, passes all your current tests, and mypy is happy with:

```py
    @functools.cached_property
    def fulldisplayname(self) -> str:
        parent: Class | Module | Clinic
        name = self.displayname
        if self.kind.new_or_init:
            if self.cls is None:
                return name
            else:
                parent = self.cls.parent
        else:
            parent = self.parent
        while isinstance(parent, (Module, Class)):
            name = f"{parent.name}.{name}"
            parent = parent.parent
        return name
```

Reply:
Or better yet:

```py
    @functools.cached_property
    def fulldisplayname(self) -> str:
        parent: Class | Module | Clinic | None
        if self.kind.new_or_init:
            parent = getattr(self.cls, "parent", None)
        else:
            parent = self.parent
        name = self.displayname
        while isinstance(parent, (Module, Class)):
            name = f"{parent.name}.{name}"
            parent = parent.parent
        return name
```